### ISSUE TITLE
P2 Composite Res: Auto-detect Aggregations on file upload

### ISSUE THREAD

**Issue by horsburgh** [hydroshare/hydroshare#2571]
Add new functionality to auto-detect Aggregations for composite resources on file upload (during resource creation and/or during resource editing).
* Geographic feature - a ".shp" file is detected
* Multidimensional - a ".nc" file is detected
* Geographic raster - a ".vrt" file is present that references one or more ".tiff" files. Or, a ".tiff" file is detected that is not referenced by a ".vrt" file.
* Time series - a ".sqlite" file is detected and is an ODM2 datbase containing time series
* Referenced time series - A ".refts.json" file is detected

NOTE: Default behavior will be auto detection. If a user doesn't want an Aggregation after it has been detected, they can remove it via right click.

**Comment by pkdash** [comment]
@horsburgh Does auto detection apply to zip file upload? A zip file possibly can be either a raster aggregation or a geo feature aggregation?

**Comment by horsburgh** [comment]
@pkdash - my first reaction is that it would be a bad idea for us to be scanning every zip file that people upload looking for potential aggregations. We don't know if there will be 5 files in a zip file or 500 files, or 5000 files. 

I've seen on some websites that when you drop a zip file on the upload space it asks - do you want to upload the zip file, or do you want to upload and expand (unzip) the file? We could potentially think about this for future work where if they choose to upload a zip file it is just uploaded, but if they choose to upload and expand, we could do the automatic detection of Aggregations. I don't think this is needed for the initial implementation.

@dtarb - do you have other thoughts about this?

**Comment by pkdash** [comment]
@horsburgh I am glad that we don't try to auto detect aggregation from zip file upload for now. May be in phase-3.

**Comment by dtarb** [comment]
I agree not to parse zip files.  However when a zip file is unzipped, we should parse the files that are added from the unzip and if we recognize them, make them into aggregations.  There are rules for this in the business logic for creating aggregations in https://docs.google.com/document/d/1Hb18tUZyG7PHSwrclRHOmcYcEJvD0i9dQSmNdVckBUc/edit.

**Comment by horsburgh** [comment]
@dtarb - this plan sounds consistent with my thinking.

**Comment by sblack-usu** [comment]
I've got auto aggregate working for Geographic Feature, Multidimensional, and Time series.

> Referenced time series - A ".refts.json" file is detected
Are we looking for extension .refts.json or json.refts?  The file you have in your demo resource is json.refts.  If it is indeed json.refts, do I need to look for that full extension, or just .refts?  I ask this because the javascript I referenced to identify these files is only looking for .refts

> Geographic raster - a ".vrt" file is present that references one or more ".tiff" files. Or, a ".tiff" file is detected that is not referenced by a ".vrt" file.
I think a conversation about geographic raster would be helpful.  The javascript I referenced is only looking for .tif files (not .tiff) and I'm not aware of a check for .vrt. 

Here is the code I'm looking referencing.
https://github.com/hydroshare/hydroshare/blob/develop/theme/static/js/hs_file_browser.js#L155

**Comment by horsburgh** [comment]
@sblack-usu - it would be good to ask for a definitive answer on the referenced time series file from the BYU group. My preference would be ".refts.json" such that ".json" (which is the actual file type) is the final extension. @pkdash gave me the file that was in my demo, so I don't even know where it came from.

With regard to the geographic raster, your quoted text is accurate and is what should be implemented. I think @pkdash is aware that the javascript testing for geographic rasters is currently incomplete and isn't detecting ".vrt" files correctly (but I think it should).

**Comment by horsburgh** [comment]
Sorry - meant to tag the BYU guys in the above comment @danames and @zhiyuli - can you chime in on the question that @sblack-usu had about referenced time series files?

**Comment by sblack-usu** [comment]
Currently, hydroshare does not allow you to set file.refts.json as a referenced time series.  I tried to aggregate a .refts.json in hydroshare and was not able to.  Setting a .json.refts file to a time series works.

I'll move forward with the Geographic raster including both .tif as well as .tiff.

Is there a reason the current system pulls the contents out of an uploaded folder?  

I also don't understand how a vrt file could reference two tif files with the current system.  The code only allows one file to be set as an aggregation at a time.  (You can't select two tif files, right click and set it a a geographic raster)

**Comment by pkdash** [comment]
@horsburgh and @sblack-usu 

Ref time series: The filename for this needs end with 'json.refts'. This was a decision made by BYU and CUAHSI. I remember there was a big discussion on this  in one of the weekly HydroShare meetings. 

Geographic Raster: Currently for raster that consists of multiple tif files there has to be a vrt file also and all of them need to be uploaded as a zip file. The additional functionality that we want is that once a folder has multiple tif files and the corresponding vrt file we should be able set that folder to raster aggregation. We should also allow the user to select the vrt file and set it to aggregation. 

But auto detection for aggregation creation may be we should do raster aggregation at the last.

**Comment by horsburgh** [comment]
For reference time series - If we ended up at ".json.refts" then we go with that. The BYU guys should confirm.

I think the folder operation where you have two .tif files and a .vrt file in a folder and then right click on the folder to set it as an aggregation already works.

I guess I don't see why if I upload two tif files and a vrt and right click on the vrt I can't convert to a geographic raster aggregation. This same type of functionality works for the geographic feature (shapefile) without requiring pre-zipping.

**Comment by sblack-usu** [comment]
I was not able to convert an unzipped folder with two .tif files and a .vrt file to an aggregation.  I can look into what geographic feature is doing and see if I can bring this functionality to the raster.

I think I need a better idea of how this is going to be used.  The zipped files helps to keep things clean and separate.  The current file uploading implementation doesn't respect folders and brings all files up to the root level.  I could see this causing a lot of confusion and bugs if someone were to upload a folder that's organized with mutliplie aggregations in separate folders.

**Comment by horsburgh** [comment]
On park if I create a folder, and upload into it two .tif files and a .vrt that references them and then right click on the folder and covert to a geographic raster aggregation, it works for me. When I download the aggregation and look at the resource map document, both .tif files and the .vrt are part of the aggregation.  

However - Yes - if i drag and drop a folder containing those three files onto the file browser, it gets rid of the containing folder, which is a real issue.

**Comment by pkdash** [comment]
@horsburgh 

> I guess I don't see why if I upload two tif files and a vrt and right click on the vrt I can't convert to a geographic raster aggregation. This same type of functionality works for the geographic feature (shapefile) without requiring pre-zipping.

The reason is how the respective resource types work. When we implemented each aggregation type we basically made the metadata extraction and file validation exactly how they work in the corresponding resource type. This was a decision made by @dtarb and me when we were considering the concept of composite resource and the reason for that is to use the same file validation code for both resource type and aggregation. It is probably time now to make the aggregation creation more flexible than what's possible with the resource type.

In implementing aggregation creation folder, the file validation logic in some cases probably have diverged from the resource type file validation as resource types have no concept of a folder.

**Comment by sblack-usu** [comment]
I was able to make some changes that preserves the folder structure of an uploaded folder.  It's a bit messy in the code right now, but at least proof that it can be done.  I'll clean it up before committing.

**Comment by sblack-usu** [comment]
@pkdash If you get a chance, I'd love some feedback on the auto-aggregation and folder path work I've done.  I'm going to be doing some refactoring on Monday.  The code is functional but it feels a bit rough around the edges.  If anything stands out to you, I'd appreciate the direction you can give.

https://github.com/hydroshare/hydroshare/compare/2502-composite-resource-phase-2...2571-auto-aggregate

**Comment by dtarb** [comment]
Responding to the comment about having to zip tiff files and a .vrt file before converting to an aggregation, this was a decision made to re-use existing raster resource type code, with the expectation that later it could be generalized as we advance functionality incrementally.  We cannot do everything at once.  Just clicking on a .vrt file without having the files zipped, requires parsing the .vrt file to identify the associated files and check if they are all there, making the activity a bit more involved.

**Comment by zhiyuli** [comment]
@dtarb @horsburgh @pkdash @sblack-usu 

The "refts.json" extension makes more sense to me. Actually I vaguely remember we discussed this before and agreed on using "refts.json" as once user downloads this file, mainstream text editor can recognize it is a JSON file.

@danames 
You still recall this?

**Comment by zhiyuli** [comment]
@dtarb @horsburgh @pkdash @sblack-usu @danames 
Our time series app developer Matt just confirmed we have been using "refts.json" as filename extension for Refts Res Type.

**Comment by pkdash** [comment]
@sblack-usu Had a quick look at your auto aggregation creation work. In reference to the new code for creating aggregations as part of uploading the files to resource, I am wondering if we need to create aggregations both from folders as well as files. I think we can just create aggregation from files only. Otherwise if we first create aggregation from a folder then try to create an aggregation from a file which may be in that folder, the aggregation creation from the file will fail. Also we should be auto generating aggregations only in case of CompositeResource.

**Comment by sblack-usu** [comment]
So, I'm looking for aggregations in the folder first because of geographic rasters in the case where we have a vrt file referencing multiple tif files.  Looking at a vrt file directly currently does not work as the current implementation expects a folder.  If we were able to update it to look for a vrt, then we'd also have to force some type of order of which aggregation types to set first (To avoid creating an aggregation off of a tif and then later trying to create an aggregation off of a vrt that references the tif). 

FWIW, I've had success manually testing this as is.  It basically mimics the rules set by the ui for setting aggregations.  We may want to consider looking for files only just to be consistent across all aggregation types, but that would also require some refactoring to change  how we're currently setting aggregations.

Yes, I've got a TODO to do auto aggregate only with CompositeResources, thanks for reminder.
